Add AdditionalData node to the AuthoriseResponse

lbailly opened the pull request:
Add Alias / AliasType fields to AdditionalData to read values returned by an authorisation request

lbailly committed:
Add AdditionalData node to the AuthoriseResponse struct in order to get more information from Adyen depending on the setting of the account
Add Alias / AliasType fields to AdditionalData to read values returned by an authorisation request

lbailly committed:
Add support to the Recurring service, need to change adyen.go and environnement.go to accept other endpoint than /Payment

zhutik commented on recurring.go:
```diff
@@ -0,0 +1,39 @@
+type RecurringDetailsResult struct {
+	CreationDate string `json:"creationDate"`
+	Details      []struct {
+		RecurringDetail struct {
```
Maybe it's better to decouple RecurringDetail?
Card structure already exists, so maybe it could be used there as well?

lbailly replied:
I'll decouple RecurringDetail.
For Card, the existing struct uses int for expiryMonth and expiryYear but JSon contains string values. So the unmarshalling process throws an errror : "json: cannot unmarshal string into Go struct field Card.expiryMonth of type int"

Below an example of Json returned by Adyen :
{"creationDate":"2017-09-27T16:30:23+02:00","details":[{"RecurringDetail":{"acquirer":"TestPmmAcquirer","acquirerAccount":"TestPmmAcquirerAccount","additionalData":{"cardBin":"411111"},"alias":"H167852639363479","aliasType":"Default","card":{"expiryMonth":"8","expiryYear":"2018","holderName":"test","number":"1111"},"contractTypes":["RECURRING"],"creationDate":"2017-09-27T16:30:23+02:00","firstPspReference":"8815065226234138","paymentMethodVariant":"visa","recurringDetailReference":"8315065226234554","variant":"visa"}},{"RecurringDetail":{"acquirer":"TestPmmAcquirer","acquirerAccount":"TestPmmAcquirerAccount","additionalData":{"cardBin":"497794"},"alias":"F314134953496447","aliasType":"Default","card":{"expiryMonth":"8","expiryYear":"2018","holderName":"test","number":"9497"},"contractTypes":["RECURRING"],"creationDate":"2017-09-27T16:58:08+02:00","firstPspReference":"8815065242885772","paymentMethodVariant":"visagold","recurringDetailReference":"8415065242885172","variant":"visa"}},{"RecurringDetail":{"acquirer":"TestPmmAcquirer","acquirerAccount":"TestPmmAcquirerAccount","additionalData":{"cardBin":"558555"},"alias":"A829629780526225","aliasType":"Default","card":{"expiryMonth":"8","expiryYear":"2018","holderName":"test","number":"5583"},"contractTypes":["RECURRING"],"creationDate":"2017-09-28T11:07:09+02:00","firstPspReference":"8515065896290058","paymentMethodVariant":"mc","recurringDetailReference":"8315065896294334","variant":"mc"}}],"shopperReference":"346","invalidOneclickContracts":"false"}

zhutik replied:
than I guess it's better to update card structure to string type for month and year
https://docs.adyen.com/developers/api-reference/payments-api#card 
f.e 1 (January) should be 01 as shown on API reference

zhutik commented:
@lbailly thanks a lot for PR! Super useful!

I'm also created a playgroud for this library https://github.com/zhutik/adyen-api-go-example maybe you can update it as well after this one get merged? It's useful for testing and verification, so make everything works as expected.

lbailly committed:
Change Card ExpireMonth / ExpirteYear / Cvc to string type to better fit Adyen structure
Create RecurringDetail struct and add it to RecurringDetailsResult, use of existing Card struct

zhutik commented:
@lbailly 

```
recurring.go:21:6: exported type RecurringDetail should have comment or be unexported
recurring_gateway.go:11:1: comment on exported method RecurringGateway.ListRecurringDetails should be of the form "ListRecurringDetails ..."
```

2 lint errors to fix.

I've fixed unit tests on TravisCI for pull requests from forks, so now it's more useful

lbailly committed:
Fix lint issues

zhutik commented on adyen.go:
```diff
@@ -97,13 +103,13 @@ func (a *Adyen) SetCurrency(currency string) {
 	}
 
-	url := a.adyenURL(method)
+	url := a.adyenURL(service, method) + "/" + service + "/"
```
Not sure if `service` should be added here. You already passing `service` to `adyenURL`  and it should return a URL like this `https://pal-test.adyen.com/pal/servlet/Recurring/v25/listRecurringDetails` or `https://pal-test.adyen.com/pal/servlet/Payment/v25/Capture`

lbailly replied:
Good catch I should have review a bit more the code ;-)

zhutik commented on adyen.go:
```diff
@@ -97,13 +103,13 @@ func (a *Adyen) SetCurrency(currency string) {
 
 // execute request on Adyen side, transforms "requestEntity" into JSON representation
-func (a *Adyen) execute(method string, requestEntity interface{}) (*Response, error) {
+func (a *Adyen) execute(method string, service string, requestEntity interface{}) (*Response, error) {
```
Maybe for consistency, we should change an order of params? now it's `method, service` but adyenURL accept `service, method`. I think `service, method` make more sense, cause we are moving from generic to the specific method that needs to be called.

lbailly replied:
No problem, I'll push an update

lbailly committed:
Little refactoring : change order of method and service parameters, encapsulate URL logic into adyenURL method

zhutik commented on adyen.go:
```diff
@@ -71,8 +77,8 @@ func (a *Adyen) ClientURL() string {
-func (a *Adyen) adyenURL(requestType string) string {
-	return a.Credentials.Env.BaseURL(APIVersion) + "/" + requestType
+func (a *Adyen) adyenURL(service string, requestType string) string {
+	return a.Credentials.Env.BaseURL(service, APIVersion) + "/" + requestType + "/" + service + "/"
```
Seems like the issue with double `service` parameter doesn't solve. Can you please check?

zhutik commented:
Merging, will fix `adyenURL` in separate commit
